### PR TITLE
Adds the ability to hedge storage requests.

### PR DESCRIPTION
**Pull request by cyriltovena**
Hedges GCS/S3/Azure/Swift requests using this [library](https://github.com/cristalhq/hedgedhttp).

There's 2 minor caveats:
- Doesn't work for the ruler.
- For Openstack the implementation also hedge auth request. Required an upstream patch if we don't want to.

**What this PR does / why we need it**:

This allow to reduce the tail latency see paper [Tail at Scale](https://cacm.acm.org/magazines/2013/2/160173-the-tail-at-scale/fulltext) by Jeffrey Dean, Luiz André Barroso. In short: the client first sends one request, but then sends an additional request after a timeout if the previous hasn't returned an answer in the expected time. The client cancels remaining requests once the first result is received.

**Special notes for your reviewer**:

**Checklist**
- [x] Documentation added
- [x] Tests updated
- [x] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

File: pkg/storage/chunk/azure/blob_storage_client.go [hunk 1]
```diff
@@ -259,7 +305,7 @@ func (b *BlobStorage) List(ctx context.Context, prefix, delimiter string) ([]chu
 }
 
 func (b *BlobStorage) DeleteObject(ctx context.Context, blobID string) error {
-	blockBlobURL, err := b.getBlobURL(blobID)
+	blockBlobURL, err := b.getBlobURL(blobID, false)
```
Comment (dannykopping) [inline pkg/storage/chunk/azure/blob_storage_client.go:308]: Isn't this operation safe? It's only the line `313` that should not be hedged, right?

File: pkg/storage/chunk/openstack/swift_object_client.go [hunk 2]
```diff
@@ -16,16 +18,26 @@ import (
 	"github.com/cortexproject/cortex/pkg/util/log"
 
 	"github.com/grafana/loki/pkg/storage/chunk"
+	"github.com/grafana/loki/pkg/storage/chunk/hedging"
 )
 
+var defaultTransport http.RoundTripper = &http.Transport{
+	Proxy:                 http.ProxyFromEnvironment,
+	MaxIdleConnsPerHost:   512,
+	ExpectContinueTimeout: 5 * time.Second,
```
Comment (dannykopping) [inline pkg/storage/chunk/openstack/swift_object_client.go:27]: Where do these values come from?

**Review by dannykopping — CHANGES_REQUESTED** [review]
I really want this feature and I think the implementation looks solid; I've added a few small questions.

I think we should not merge this though until we have a way to deal with the following situation:

GCS/S3/whatever is experiencing a partial outage, which is causing its latency to spike. This is adding 200ms to all requests. This additional latency causes *all* of our requests to get hedged, which actually does more harm than good here since it increases tail latencies across the board ironically.

AFAICS from this PR, there is no protection against this.
I'm also not sure if any metrics have been added to see how many requests are hedged - did I miss something?

One dumb solution would be to add a config option for the maximum number of hedged requests per querier per second/minute.

File: pkg/storage/chunk/azure/blob_storage_client.go (hunk 1 above)
Comment (cyriltovena) [reply pkg/storage/chunk/azure/blob_storage_client.go:308]: Not sure what you means, only gets are hedged.

File: pkg/storage/chunk/openstack/swift_object_client.go (hunk 2 above)
Comment (cyriltovena) [reply pkg/storage/chunk/openstack/swift_object_client.go:27]: From the openstack library.

File: pkg/storage/chunk/azure/blob_storage_client.go (hunk 1 above)
Comment (dannykopping) [reply pkg/storage/chunk/azure/blob_storage_client.go:308]: Ah sorry, I see `blockBlobURL` is used by the `Delete` call; I thought this was a separate `GET` operation

File: pkg/storage/chunk/aws/s3_storage_client_test.go [hunk 3]
```diff
@@ -75,3 +81,65 @@ func TestRequestMiddleware(t *testing.T) {
...
+func Test_Hedging(t *testing.T) {
+	for _, tc := range []struct {
+		name          string
+		expectedCalls int32
+		hedgeAt       time.Duration
+		upTo          int
+		do            func(c *S3ObjectClient)
+	}{
+		{
+			"deletes are not hedged",
```
[the remark this answers is not in the document]
Comment (cyriltovena) [reply pkg/storage/chunk/aws/s3_storage_client_test.go]: I added all possible query that should not be hedged to make it more clear.

File: pkg/storage/chunk/azure/blob_storage_client.go (hunk 1 above)
Comment (cyriltovena) [reply pkg/storage/chunk/azure/blob_storage_client.go:308]: see above.

**Comment by cyriltovena** [comment]
> I really want this feature and I think the implementation looks solid; I've added a few small questions.
> 
> I think we should not merge this though until we have a way to deal with the following situation:
> 
> GCS/S3/whatever is experiencing a partial outage, which is causing its latency to spike. This is adding 200ms to all requests. This additional latency causes _all_ of our requests to get hedged, which actually does more harm than good here since it increases tail latencies across the board ironically.
> 
> AFAICS from this PR, there is no protection against this. I'm also not sure if any metrics have been added to see how many requests are hedged - did I miss something?
> 
> One dumb solution would be to add a config option for the maximum number of hedged requests per querier per second/minute.

Unfortunately all those ask require improvement of the library upstream. I guess we could do that.

If we all think we should do it, then I can look into it.

**Comment by dannykopping** [comment]
> > I really want this feature and I think the implementation looks solid; I've added a few small questions.
> > I think we should not merge this though until we have a way to deal with the following situation:
> > GCS/S3/whatever is experiencing a partial outage, which is causing its latency to spike. This is adding 200ms to all requests. This additional latency causes _all_ of our requests to get hedged, which actually does more harm than good here since it increases tail latencies across the board ironically.
> > AFAICS from this PR, there is no protection against this. I'm also not sure if any metrics have been added to see how many requests are hedged - did I miss something?
> > One dumb solution would be to add a config option for the maximum number of hedged requests per querier per second/minute.
> 
> Unfortunately all those ask require improvement of the library upstream. I guess we could do that.
> 
> If we all think we should do it, then I can look into it.

Does it have to be changed upstream? We could create a layer of indirection which both tracks all hedged requests and cancels it if there have been too many.

Having it upstream would be nice, too.

**Review by jeschkies — APPROVED** [review]

File: pkg/storage/chunk/aws/s3_storage_client_test.go [hunk 4]
```diff
@@ -91,12 +92,14 @@ func Test_Hedging(t *testing.T) {
...
 		{
-			"deletes are not hedged",
-			1,
+			"delete/put/list are not hedged",
+			3,
 			20 * time.Nanosecond,
 			10,
 			func(c *S3ObjectClient) {
 				_ = c.DeleteObject(context.Background(), "foo")
+				_, _, _ = c.List(context.Background(), "foo", "/")
```
Comment (dannykopping) [inline pkg/storage/chunk/aws/s3_storage_client_test.go:101]: I don't understand this one, sorry; why would a `List` call (which appears to issue a `GET`?) not be hedged?

Comment (cyriltovena) [reply pkg/storage/chunk/aws/s3_storage_client_test.go:101]: The point is to hedge fetching chunks, which is the what is in the path of requests latency.

List is mostly used for listing rules, or index files, or the compactor.

Comment (dannykopping) [reply pkg/storage/chunk/aws/s3_storage_client_test.go:101]: Gotcha :ok_hand:

**Comment by cyriltovena** [comment]
> > > I really want this feature and I think the implementation looks solid; I've added a few small questions.
> > > I think we should not merge this though until we have a way to deal with the following situation:
> > > GCS/S3/whatever is experiencing a partial outage, which is causing its latency to spike. This is adding 200ms to all requests. This additional latency causes _all_ of our requests to get hedged, which actually does more harm than good here since it increases tail latencies across the board ironically.
> > > AFAICS from this PR, there is no protection against this. I'm also not sure if any metrics have been added to see how many requests are hedged - did I miss something?
> > > One dumb solution would be to add a config option for the maximum number of hedged requests per querier per second/minute.
> > 
> > 
> > Unfortunately all those ask require improvement of the library upstream. I guess we could do that.
> > If we all think we should do it, then I can look into it.
> 
> Does it have to be changed upstream? We could create a layer of indirection which both tracks all hedged requests and cancels it if there have been too many.
> 
> Having it upstream would be nice, too.

The library uses the http.Roundtripper pattern:
- if you wrap after you don't know if you're a hedge request or a normal one.
- if you wrap before you can't predict if the request would be hedge or not.

It might be possible by wrapping before AND after, but that seems complex. I'll give a go if you don't hear from me on that matter means I couldn;t :)

**Review by owen-d — APPROVED** [review]
This looks good and I like @dannykopping's suggestion to limit the number of hedged requests to a certain percentage of total reqs, but I also don't want perfect to be the enemy of good and think this is beneficial enough on it's own. 

Instead of making this a per-store option, could we implement it a level higher by wrapping the storage client interface to create a hedging client? That would also allow us to expose only one hedging config block, rather than one per backend.

**Comment by cyriltovena** [comment]
FYI I asked this before submitting this PR. 

https://github.com/cristalhq/hedgedhttp/issues/17

**Comment by cyriltovena** [comment]
> Instead of making this a per-store option, could we implement it a level higher by wrapping the storage client interface to create a hedging client? That would also allow us to expose only one hedging config block, rather than one per backend.

I hesitated to do this although I realized it won't be applicable to some other backend like grpc or local. But if we think that doesn't matter I'm in for making it broader.

**Comment by dannykopping** [comment]
> Instead of making this a per-store option, could we implement it a level higher by wrapping the storage client interface to create a hedging client? That would also allow us to expose only one hedging config block, rather than one per backend.

Yeah I really like that, if possible

**Comment by cyriltovena** [comment]
> Yeah I really like that, if possible

Done ✨

**Review by dannykopping — APPROVED** [review]
Approving, we'll add the hedging rate-limiting in a follow-up PR

**Pull request merged**